cmd/promtool/backfill.go

```diff
@@ -0,0 +1,136 @@
+// Copyright 2020 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"math"
+	"os"
+	"time"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/pkg/errors"
+	"github.com/prometheus/prometheus/pkg/labels"
+	"github.com/prometheus/prometheus/pkg/textparse"
+	"github.com/prometheus/prometheus/tsdb"
+	tsdb_errors "github.com/prometheus/prometheus/tsdb/errors"
+	"github.com/prometheus/prometheus/tsdb/importer/openmetrics"
+)
+
+var merr tsdb_errors.MultiError
+
+func getMinAndMaxTimestamps(p textparse.Parser) (int64, int64, error) {
+	var (
+		entry textparse.Entry
+		err   error
+	)
+	maxt, mint := int64(math.MinInt64), int64(math.MaxInt64)
+	for {
+		entry, err = p.Next()
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return maxt, mint, errors.Wrap(err, "parse")
+		}
+		if entry != textparse.EntrySeries {
+			continue
+		}
+		l := labels.Labels{}
+		p.Metric(&l)
+		_, ts, _ := p.Series()
+		if *ts > maxt {
+			maxt = *ts
+		}
+		if *ts < mint {
+			mint = *ts
+		}
+		if ts == nil {
+			return maxt, mint, errors.Errorf("Expected timestamp for series %v, got none.", l.String())
+		}
+	}
+	return maxt, mint, nil
+}
+func createBlocks(input *os.File, mint, maxt *int64, outputDir string) error {
+	logger := log.NewNopLogger()
+	var offset int64 = 2 * time.Hour.Milliseconds()
+	(*mint) = offset * ((*mint) / offset)
+	for t := (*mint); t <= (*maxt); t = t + offset {
+		w, errw := tsdb.NewBlockWriter(log.NewNopLogger(), outputDir, offset)
+		if errw != nil {
+			os.RemoveAll(outputDir)
```

Comment:
Why are you doing this only on this code path?

Reply:
If there is an error while creating the blocks, remove the blocks and return error.

Reply:
But why only this error path?

Reply:
Perhaps this is a candidate for a `defer`'d call.

Reply:
> But why only this error path?
Because only if there error I want to delete the outputDir, but yes I should put it in defer call untill all error checks are completed.

Reply:
IIUC, putting it in a defer'd call would mean that it will be called after the neighbouring functions complete. But we only want to clean the directory when createBlocks fails. So, I have moved it here 
```
if errCb := createBlocks(input, mint, maxt, outputDir); errCb != nil {
    os.RemoveAll(outputDir)
    return errors.Wrap(errCb, "block creation")
}
```